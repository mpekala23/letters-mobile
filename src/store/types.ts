import { UserState } from "./User/UserTypes";
import { NotifState } from "./Notif/NotifTypes";
import { ContactState } from "./Contact/ContactTypes";

export interface AppState {
  user: UserState;
  notif: NotifState;
  contact: ContactState;
}
