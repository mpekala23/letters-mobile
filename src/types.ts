// TODO: make this typing better
// probably want a specific type for state abbrevs
export interface Facility {
  name: string;
  type: PrisonTypes;
  address: string;
  city: string;
  state: string;
  postal: string;
  phone: string;
}

export enum ProfilePicTypes {
  Topbar = 'Topbar',
  Contact = 'Contact',
  SingleContact = 'SingleContact',
}

export enum Storage {
  RememberToken = 'Ameelio-Token',
}

export enum PrisonTypes {
  State = 'State',
  Federal = 'Federal',
  Jail = 'County',
}

export enum SupportFAQTypes {
  DeleteLetter = 'DeleteLetter',
  NotArrived = 'NotArrived',
  WrongReturnAddress = 'WrongReturnAddress',
  WrongMailingAddress = 'WrongMailingAddress',
  TrackingNumber = 'TrackingNumber',
  TrackingError = 'TrackingError',
  TalkToAmeelio = 'TalkToAmeelio',
}

export enum LetterStatus {
  Draft = 'Draft',
  Created = 'Created',
  Mailed = 'Mailed',
  InTransit = 'In Transit',
  InLocalArea = 'In Local Area',
  OutForDelivery = 'Out for Delivery',
  Delivered = 'Delivered',
}

export enum LetterTypes {
  Postcard = 'postcard',
  Letter = 'letter',
}

export enum DeliveryReportTypes {
  received = 'received',
  unsure = 'unsure',
  notYetReceived = 'notYetReceived',
  haveNotAsked = 'haveNotAsked',
  haveNotReceived = 'haveNotReceived',
}

export interface Letter {
  type: LetterTypes;
  status: LetterStatus;
  isDraft: boolean;
  recipientId: number;
  content: string;
  photo?: Photo;
  letterId?: number; // TODO: Once we have more info on this field and lob, use this more
  expectedDeliveryDate?: Date;
  dateCreated?: Date;
  trackingEvents?: LetterTrackingEvent[];
}

export interface LetterTrackingEvent {
  id: number;
  name: string;
  location: ZipcodeInfo;
  date: Date;
}

export type TopbarRouteAction = {
  enabled: boolean;
  text: string;
  action: () => void | Promise<void>;
  blocking?: boolean;
};

export interface ZipcodeInfo {
  zip: string;
  city: string;
  state: string;
  lat?: number;
  long?: number;
}

export interface Photo {
  type?: 'image' | string;
  width?: number;
  height?: number;
  uri: string;
}
