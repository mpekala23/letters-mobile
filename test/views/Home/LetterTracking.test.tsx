import * as React from 'react';
import { LetterTrackingScreen } from '@views';
import { render, toJSON, fireEvent } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { LetterTypes, LetterStatus } from 'types';

const mockStore = configureStore([]);

jest.mock('moment', () => () => ({
  format: () => 'Jun 30',
}));

const setup = (letterOverrides = {}) => {
  const navigation = { navigate: jest.fn(), addListener: jest.fn() };
  const letter = {
    type: LetterTypes.Postcard,
    status: LetterStatus.Mailed,
    isDraft: true,
    recipientId: 8,
    content: "I'm trying out this new service called Ameelio...",
    letterId: 1,
    expectedDeliveryDate: '2019-06-30',
    trackingEvents: [
      {
        id: 1,
        name: LetterStatus.Mailed,
        location: '20002',
        date: '2019-07-12T15:51:41.000Z',
      },
    ],
    ...letterOverrides,
  };

  const initialLetterState = {
    composing: {},
    active: letter,
    existing: [],
  };

  const store = mockStore({
    letter: initialLetterState,
    notif: { currentNotif: null },
  });

  const StoreProvider = ({ children }: { children: JSX.Element }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return {
    navigation,
    store,
    ...render(<LetterTrackingScreen navigation={navigation} />, {
      wrapper: StoreProvider,
    }),
  };
};

describe('Letter Tracking Screen', () => {
  it('should match snapshot', () => {
    const { container } = setup(<LetterTrackingScreen />);
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it('should display formatted delivery date', () => {
    const { getByTestId } = setup();
    expect(getByTestId('deliveryDate').props.children).toBe('Jun 30');
  });

  it('should display letter content', () => {
    const { getByText } = setup();
    expect(
      getByText("I'm trying out this new service called Ameelio...")
    ).toBeDefined();
  });

  it('should load values for letters from the redux store', () => {
    const { getByText } = setup({
      type: LetterTypes.Postcard,
      status: LetterStatus.InTransit,
      isDraft: true,
      recipientId: 8,
      content: 'Redux Letter 1',
      letterId: 2,
      expectedDeliveryDate: '2019-06-30',
    });
    expect(getByText('Redux Letter 1').props.children).toBe('Redux Letter 1');
  });

  it('should navigate to Support FAQ screen when button is pressed', async () => {
    const { navigation, getByText } = setup();
    fireEvent.press(getByText('I need help'));
    expect(navigation.navigate).toHaveBeenCalledWith('SupportFAQ');
  });
});
