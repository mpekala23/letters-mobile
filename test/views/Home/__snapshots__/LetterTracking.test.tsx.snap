// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Letter Tracking Screen should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollView
    keyboardShouldPersistTaps="handled"
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "column",
        "padding": 16,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "paddingBottom": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Poppins-Bold",
              },
              Object {
                "color": "#282828",
                "fontSize": 21,
                "paddingBottom": 4,
              },
            ]
          }
        >
          Letter tracking
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Poppins-Bold",
              },
              Object {
                "color": "#282828",
                "fontSize": 14,
              },
            ]
          }
        >
          Your letter is estimated to arrive on
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Poppins-Bold",
              },
              Object {
                "color": "#282828",
                "fontSize": 14,
              },
            ]
          }
          testID="deliveryDate"
        >
          Jun 30
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#F2F2F2",
            "height": 2,
            "marginVertical": 8,
            "width": "100%",
          }
        }
      />
      <View
        style={
          Object {
            "paddingTop": 24,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F2F2F2",
              "height": 0,
              "marginLeft": 14,
              "marginTop": 40,
              "position": "absolute",
              "width": 7,
            }
          }
        />
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "paddingBottom": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#CFF3DD",
                  "borderRadius": 50,
                  "height": 28,
                  "marginLeft": 4,
                  "marginRight": 24,
                  "width": 28,
                }
              }
              testID="trackerCircle"
            />
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Bold",
                    },
                    Object {
                      "color": "#282828",
                      "fontSize": 16,
                      "paddingBottom": 6,
                    },
                  ]
                }
              >
                Mailed
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Bold",
                    },
                    Object {
                      "color": "#393939",
                      "fontSize": 14,
                      "paddingBottom": 4,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-end",
                  "flexDirection": "column",
                  "marginLeft": "auto",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Regular",
                    },
                    Object {
                      "color": "#393939",
                      "fontSize": 14,
                      "paddingBottom": 8,
                    },
                  ]
                }
                testID="dateFormatted"
              >
                Jun 30
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins-Regular",
                    },
                    Object {
                      "color": "#393939",
                      "fontSize": 14,
                      "paddingBottom": 4,
                    },
                  ]
                }
              >
                Jun 30
              </Text>
            </View>
          </View>
        </View>
      </View>
      <TouchableOpacity
        activeOpacity={0.7}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderColor": "#FF7171",
              "borderRadius": 8,
              "borderWidth": 2,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginVertical": 5,
            },
            Object {},
            Object {
              "alignSelf": "center",
              "height": 50,
              "marginBottom": 16,
              "width": "50%",
            },
            Object {},
            Object {
              "elevation": 2,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.22,
            },
          ]
        }
        testID="clickable"
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Poppins-Medium",
              },
              Object {
                "color": "#FF7171",
                "fontSize": 18,
                "fontWeight": "700",
                "padding": 6,
              },
              Object {},
              Array [
                Object {
                  "fontFamily": "Poppins-Bold",
                },
                Object {
                  "fontSize": 14,
                },
              ],
              Object {},
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          I need help
        </Text>
      </TouchableOpacity>
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Poppins-Bold",
          },
          Object {
            "color": "#282828",
            "fontSize": 21,
            "paddingBottom": 4,
          },
        ]
      }
    >
      Letter content
    </Text>
    <Text
      style={
        Object {
          "fontSize": 15,
        }
      }
    >
      I'm trying out this new service called Ameelio...
    </Text>
    <View
      style={
        Object {
          "height": 40,
        }
      }
    />
  </ScrollView>
</View>
`;
